ql/src/java/org/apache/hadoop/hive/ql/udf/ptf/ValueBoundaryScanner.java

```diff
@@ -664,6 +632,87 @@ public Object computeValue(Object row) throws HiveException {
    */
   public abstract boolean isEqual(Object v1, Object v2);
 
+  protected int binarySearchBack(int r, PTFPartition p, Object sortKey, int amt,
+      Order order) throws HiveException {
+    boolean isOrderDesc = order.equals(Order.DESC);
+    Object rowVal = sortKey;
+
+    int rMin = -1;  // tracks lowest possible number fulfilling the range requirement
+    int rMax = r; // tracks highest possible number fulfilling the range requirement
+
+    boolean isDistanceGreater = true;
+    while (true) {
+      if (rMin == rMax) {
+        return rMin;
+      }
+      isDistanceGreater =
+          isDistanceGreater(isOrderDesc ? rowVal : sortKey, isOrderDesc ? sortKey : rowVal, amt);
+      if (!isDistanceGreater && r == 0) {
+        return 0;
+      }
+      if (isDistanceGreater) {
+        rMin = r + 1;
+      } else {
+        rMax = r;
+      }
+      r = rMin + (rMax - rMin) / 2;
```

Comment:
"r" is the moving variable in this algorithm that tracks the currently investigated row, and we already decided to rename this to "rowId" in the method signature, so this will be rowId (as there is only 1 variable tracking this)

Reply:
Right, makes sense